### DISCUSSION TITLE
IDE for building GUIs

### DISCUSSION THREAD

**Discussion by Wajinn** [vlang/v#12985] [General]
Seems like many are using Visual Studio Code or maybe Notepad++, but these aren't arguably the best for building GUIs.  Wondering if there are any pure vlang IDEs in the works.

**Comment by ghost** [comment]
Wow too BIG DREAM lol
yeah, maybe someday lol

**Reply by Wajinn** [reply]
Nice to see your reply kasha, new to vlang, so learning lots of new things.  I was thinking that drag and drop of components in a GUI creator would be so sweet.  

I happen to recently come across an interesting simple form creator.  In their language, they also do the declarative style, but it can be sometimes difficult to type out what you want versus drag and drop.

Something simple and useful could be a forms creator (https://github.com/planetsizecpu/forms), where you drag and drop, and code is more automatically created.
(https://raw.githubusercontent.com/planetsizecpu/forms/master/formstest01.gif)

Thinking that such a concept would be great for vlang users trying to create GUIs.  Something like this can also work along existing editors like Visual Studio Code or Notepad++ (for example), where you customize them to call up the form creator, create the GUI, and then when done copy and paste the code back into the editor.  In some other discussion forums, saw that its an issue where people ( interested hobbyist/non-professionals) don't know the best ways to proceed with UI creation in vlang.

**Comment by ghost** [comment]
First of all, try to actually use the V-UI by yourself and create a UI program.
You will notice three things.
1) You can create layouts very easily with just the V-UI, so there is no need for extra applications to build such layouts.
2) V-UI itself is so simple that creating such an application would be far more time-consuming.
3) This is the most important thing, but there are still many widgets missing in V-UI, so we are still far from needing such a big program.

**Reply by Wajinn** [reply]
Don't get me wrong, like vlang and think the cross-platform UI project is very important.  V-UI is very much appreciated.  Did already use V-UI and gone through various examples.

With that stated, from the perspective of an "outsider" or casual, there are some inherent issues that make it difficult to use.  Possibly, where those that are advanced/professional programmers might not realize or are in their own bubble so can't see.

1) Installation of V-UI

The way that V-UI is setup, the person should have Git.  Telling people to "V up", will already unintentionally probably drop scores of potential users, and "V install ui" will usually fail afterwards (if no Git).  The downloadable releases of V, will work by itself, but V-UI is not included and a separate project from that.  Thus can be a point of confusion.

Yes, you can import ui and gg, but that is already diverging from any instructions on (github.com/vlang/ui) for how to use and there is no documentation for clarity. The documentation (https://github.com/vlang/v/blob/master/doc/docs.md), doesn't really even mention V-UI.  There is one ebook by Packt to teach V programming, an it barely mentions V-UI, which is just to show the link to download it (on part of a page).

2) User Friendly & Practical

The link to the simple GUI designer that I showed, has an executable file and one source file.  Thus it would be comparatively easier for users to incorporate it with editors they might be using, like Visual Studio Code or Notepad++.  If they need to build a GUI, there is also the option to run the GUI designer that helps build the code, and then paste into their choice of editor.

There is probably nothing more user friendly than "drag and drop" with automatic/semi-automatic GUI code creation.  I'm not saying anything negative about the declarative style (and V-UI has a decent system), as it can be preferred in various circumstances.  Coming from the perspective that users might want to quickly play around with various looks of a potential GUI, drag and drop can be initially faster when they are not clear as to what they want and testing ideas.

Just suggesting that an additional simple drag and drop GUI designer might be worth giving strong consideration to.  Particularly when other programming languages in a similar category already have such.  It would seem that it would help reduce the need for documentation as well, as GUI creation would be even more self explanatory.

**Reply by ghost** [reply]
Don't get me wrong, I don't blame you nor stop YOU to work and contribute to V-UI.
Yes, If there were such a extra GUI builder It might be a little bit useful to build GUI only for programing beginners.
So YOU can contribute such a extra application.

It is too cost effective.

Personally I would be appreciated if you would contrite to V-UI itself with basic Widgets rather than what most of us does not need.

Anyway, take a look at the V-UI source code for yourself.
You'll see right away that it's not the time to be thinking about a GUI builder lol

And as you mentioned, what we need first is such a instructions not a GUI builder.
Then how about starting to provide such a documents?
I think it's much more realistic, feasible, and will be appreciated by many people.

**Reply by EmmaTheMartian** [reply]
I love the idea of a GUI builder, I hate them personally, because it takes out the fun in the programming side, but V-UI is just nesting mayhem, and the window creation process takes forever. It's also quite hard to read V-UI's source code, and code that is wrote using V-UI, at times, I can't even read my *own* code that uses V-UI. Due to those reasons, a GUI builder for V-UI would make the UI creation process significantly easier.

Perhaps allowing people to write an XML file, or something similar for making the GUI creation would be even better than a GUI builder. That would beat the problem of needing to make a giant piece of software for a small need, rather, just an XML parser (Or a parser for some other language) and transpile that to V-UI-compatible code, or just straight up generate a UI at runtime.

**Reply by JalonSolov** [reply]
Please, no XML.  I would even go for YAML over XML, and I despise YAML.  V at least has a TOML parser now, so that would likely be the best option if people don't want to use something sane like JSON.

**Reply by EmmaTheMartian** [reply]
I was using XML as an example, I personally hate it. YAML is my go-to, although because V has no YAML parser, I just use TOML :joy:

**Reply by Wajinn** [reply]
> Please, no XML. I would even go for YAML over XML, and I despise YAML. V at least has a TOML parser now, so that would likely be the best option if people don't want to use something sane like JSON.

Agree, JSON maintains one's sanity.  Dislike XML too.

**Reply by Wajinn** [reply]
> I love the idea of a GUI builder, I hate them personally, because it takes out the fun in the programming side, but V-UI is just nesting mayhem, and the window creation process takes forever. It's also quite hard to read V-UI's source code, and code that is wrote using V-UI, at times, I can't even read my _own_ code that uses V-UI. Due to those reasons, a GUI builder for V-UI would make the UI creation process significantly easier.
> 
Agree with you up until XML was mentioned.

**Reply by ghost** [reply]
> I love the idea of a GUI builder, I hate them personally, because it takes out the fun in the programming side, but V-UI is just nesting mayhem, and the window creation process takes forever. It's also quite hard to read V-UI's source code, and code that is wrote using V-UI, at times, I can't even read my _own_ code that uses V-UI. Due to those reasons, a GUI builder for V-UI would make the UI creation process significantly easier.
> 
> Perhaps allowing people to write an XML file, or something similar for making the GUI creation would be even better than a GUI builder. That would beat the problem of needing to make a giant piece of software for a small need, rather, just an XML parser (Or a parser for some other language) and transpile that to V-UI-compatible code, or just straight up generate a UI at runtime.


The declarative style is Alex's intention.

I like Delphi.
I also use GUI builder in Android.
What are you confused about?
I said that the V-UI GUI builder was too big a dream, that it was too difficult now, and that good to become someday.
I said it would be much faster to write the UI code directly than to spend the time and effort to create a GUI builder, and that it would be easier to write than to use a builder.
Read carefully.
If you want to make a GUI builder, you can.
I said I'm not against it.

Besides, this is the V language page, so please write your complaints about V-UI https://github.com/vlang/ui

If it's easier and quicker, I'll take it.
I don't do programming as a hobby, I do it as part of my job.
I'm saying that it's easier and faster to learn a little and write code directly than to use a GUI builder.
If you think it's hard to write code directly, you're nothing but lazy.
I hate people who are so lazy that they force their demands on others, saying, "I want this, I want that, I don't want to do it myself because it's too much work, someone else do it for me."
I hate people who blurt out unrealistic dreams, but don't try to do any of them on their own, but try to get others to do them, and try to persuade others in a long and persistent manner without making any sacrifices on their own.

So, it's okay to want it, but let's not be pushy about it.

Try to study hard or contribute for you dream.
Good luck.

**Comment by ylluminate** [comment]
There was discussion of a UI/GUI IDE back a year or more ago, yes. If I recall correctly there was some thoughts about modeling something after the [Delphi/FreePascal IDE: Lazarus](https://www.lazarus-ide.org/). There were some other ideas and discussions as well, but obviously everyone has been busy grinding away in other areas for the time being.

I would definitely encourage you, if interested, in starting up something like this. You are absolutely right about a GUI designer being invaluable and while things are very much in flux, if this is comprehended, could offer a nice test platform in addition to the ultimately useful goals of productivity.

**Reply by Wajinn** [reply]
I'm just a curious person passing through and looking at something new, not an advanced professional programmer, so wouldn't be the one to implement such.  It maybe should be kept in mind that not everyone has the background, inclination, time, or will to be a developer.  I haven't even made up my mind to more thoroughly use vlang versus other languages.  Hopefully people of greater or developer skill level already exist in the vlang community.  If not, that would be surprising.  Hoping that nobody is overly upset at a new curious user making suggestions.

There is also something to be said that there are going to be plain users of projects versus everyone that touches the project is also advanced in programming or a developer.  There should be much more highly skilled vlang programmers (probably with backgrounds in other programming languages), who hopefully are interested.  

As was touched on by ylluminate, vlang.io states this, "Coming soon: a Delphi-like visual editor for building native GUI apps".  

I don't know how many years that might have been up there.  So the idea of at least a simple GUI designer shouldn't be a shock of any kind.  Also, something like Delphi/Lazarus is quite heavy.  Even to download and setup is quite extensive and serious.  What I linked to in a previous post is much more lightweight and quick, and maybe something to go along with other already existing lightweight editors like Visual Studio Code and Notepad++... or even just Notepad.  At the end of the day, its just a suggestion.  If vlang hopes to more fully compete against Go or the likes of Free Pascal/Lazarus, its probably something to strongly consider.

**Comment by ghost** [comment]
A full-fledged GUI builder like Delphi, VisualStudio, or AndroidStudio would certainly be in demand.
But there is no demand for such half-baked GUI builders.
If you are a beginner who cannot program without a GUI builder, you should use an IDE made by a company such as Delphi, VisualStudio or AndroidStudio.

For a minor language like V, which is open-source and developed by a small number of developers who volunteer their time for free, it is impossible to make a corporate IDE, and unfortunately, no one is willing to spend time on such a costly and ineffective GUI builder, even though there is little demand for it.

**Reply by Wajinn** [reply]
Forgive my ignorance, but I would think that potential vlang users would be even more interested in using it to create GUIs, to include cross-platform programs.  Many traditional programming languages neglect(ed) GUI frameworks and libraries, treating them more of an afterthought.  Vlang appears to have made this a fundamental consideration of the language from the beginning.

It also seems contradictory that beginners would be more satisfied with Visual Studio or Delphi, because these are heavy and involved IDEs (including expensive), which are also targeted towards the enterprise and professional programmers.  Even Lazarus (Free Pascal), isn't something so easy and quick to play with. More "light" languages and tools from Python (though some disagree about them), Lua, or AutoHotkey (popular Windows scripting language that also has Adventure IDE/AutoGUI (https://sourceforge.net/projects/autogui/)) offer open source alternatives that are much easier for beginners and casuals to create useful things quickly.

**Reply by ghost** [reply]
If you're going to go that far, let me be clear: V is certainly an easy-to-use language, but it is currently buggy and has strong quirks. There are not many users, so there is not much information to refer to.
It is not a language that can be handled by a beginner who cannot write a program without using a GUI builder, or a beginner who cannot think and write a program by himself using only existing documents and samples.
If you are a beginner at that level, you should learn programming first in a language like Go, which is developed by a company, has a lot of users, and has a lot of resources on the net for beginners.
A language like V, which is buggy and has few users, is for developers who are at least capable of solving problems on their own.

**Reply by ghost** [reply]
And your perception is fundamentally wrong.
Companies are in the business of developing IDEs.Whether users are professionals or amateurs, they are developing for business. They pay their employees to develop their products.That's why they can provide generous support and rich products.
Open source is developed by volunteers for free. They are not being paid to develop for the users.
So, why would they want to give free and time consuming support to beginners like you who don't know anything, who don't want to make any effort?
But of course no one stop you to pray for that someone else will think the same way you do and will be willing to contribute by making such a half-baked GUI builder without payment.

**Comment by JalonSolov** [comment]
Hmm... this thread seems to be getting a bit negative, to me.

It is true that V is still young, and doesn't have support for all the nice things that other languages have.  This doesn't mean we should be telling people "do it yourself - don't bother us."

V, just like other open source projects, relies on the contributions of users, be that code, or simply ideas.  Please don't stop contributing, but also be aware that _because_ it is an open source project, things may take a while to get done - if they are done at all.  It will rely on someone volunteering their time to work on it.

**Reply by Wajinn** [reply]
Agree, that it seems to be getting a bit contentious, and my intentions are not to drag it to that.  Negativity and contentiousness can cause people to feel unwelcome, or if they do create something, to not share.  I would hope that the V language can embrace all those who are interested in it, from newbies and casuals to professionals.

And, as I familiarize myself more with the V language, I may make contributions in various different ways.  I think that V is a worthy language, that has some fantastic features that distinguishes it from others.

**Comment by ghost** [comment]
No, not negative. If we want to create a cross-platform GUI application, and want to do it in an easy and lightweight environment, we don't have many options. That's how I ended up with V.
But the reality is that V is still in WIP,  so we need to learn hard, and so we can get cross-platform apps faster by learning the existing code rather than waiting for such a rich tools to appear.
If one can't do that, it would be much more constructive to use another language that is more mature.

**Reply by Wajinn** [reply]
>kasha: "...so we need to learn hard..."

We can at least agree on this point.  You have my respect for your contributions and being passionate about V.

**Reply by ghost** [reply]
You can feel free to join discord for any help no matter what level programmer.
It is the only place people can get detailed and latest knowledge about V now.
